### ISSUE TITLE
Argument type is partially unknown for map

### ISSUE THREAD

**Issue by DanielPechersky** [microsoft/pylance-release#1400]
<h3> Environment data </h3>

-   Language Server version: 2021.6.0
-   OS and version: darwin x64
-   Python version (and distribution if applicable, e.g. Anaconda): python3.10.0b2
-   python.analysis.indexing: undefined
-   python.analysis.typeCheckingMode: strict

<h3> Expected behaviour </h3>

No errors, type checker should know that map(list, list[list[X]]) -> list[list[X]]

<h3> Actual behaviour </h3>

Argument type is partially unknown for map(list, lst)

<h3> Code Snippet / Additional information </h3>

```python
lst: list[list[int]] = [[1,2,3], [4]]
lst2: list[list[int]] = list(map(list, lst))
#                            ^^^^^^^^^^^^^^
# Argument type is partially unknown
#   Argument corresponds to parameter "iterable" in function "__init__"
#   Argument type is "map[list[Unknown]]" Pylance(reportUnknownArgumentType)
```

**Comment by erictraut** [comment]
This behavior is a known limitation of [bidirectional type inference](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#bidirectional-type-inference-expected-types). In the first case (with variable `lst`), the type engine is able to use the type annotation on the LHS (left-hand side) of the assignment to inform the inferred type of the RHS expression. In the second case (with variable `lst2`), this cannot be done because bidirectional type inference cannot be applied to argument expressions used within a call expression. Therefore, the inferred type of `map(list, lst)` is `list[unknown]`.

In cases like this, you have several options:

1. Explicitly specialize the `list` argument to the `map` call:

```python
lst2 = list(map(list[int], lst))
```

2. Use a temporary variable with an annotation:

```python
lst2_temp: map[list[int]] = map(list, lst)
lst2 = list(lst2_temp)
```

3. Use a cast (not a great option because it overrides type checker logic):

```python
lst2 = list(cast(map[list[int]], map(list, lst)))
```

4. Use a "# type: ignore" comment to suppress the error (also not a great option).

```python
lst2: list[list[int]] = list(map(list, lst)) # type: ignore
```

**Comment by DanielPechersky** [comment]
Alright, that sounds good. Specializing list sounds like the best approach here.

**Comment by DanielPechersky** [comment]
Actually, why doesn't specialization work for this case?
```python
lst = [[1,2,3],[4]]
tup: tuple[tuple[int]] = tuple(map(tuple[int, ...], lst))
#                              ^^^^^^^^^^^^^^^^^^^^^^^^^
# No overloads for "__init__" match the provided arguments
#   Argument types: (Type[tuple[int, ...]], list[list[int]]) Pylance(reportGeneralTypeIssues)
```

**Comment by erictraut** [comment]
That's a bug! Thanks for reporting it. This will be fixed in the next release.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.6.1, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202161-9-june-2021
